Lib/uuid.py

```diff
@@ -756,6 +757,71 @@ def uuid5(namespace, name):
     int_uuid_5 |= _RFC_4122_VERSION_5_FLAGS
     return UUID._from_int(int_uuid_5)
 
+_last_timestamp_v7 = None
+_last_counter_v7 = 0  # 42-bit counter
+
+def uuid7():
+    """Generate a UUID from a Unix timestamp in milliseconds and random bits.
+
+    UUIDv7 objects feature monotonicity within a millisecond.
+    """
+    # --- 48 ---   -- 4 --   --- 12 ---   -- 2 --   --- 30 ---   - 32 -
+    # unix_ts_ms | version | counter_hi | variant | counter_lo | random
+    #
+    # 'counter = counter_hi | counter_lo' is a 42-bit counter constructed
+    # with Method 1 of RFC 9562, §6.2, and its MSB is set to 0.
+    #
+    # 'random' is a 32-bit random value regenerated for every new UUID.
+    #
+    # If multiple UUIDs are generated within the same millisecond, the LSB
+    # of 'counter' is incremented by 1. When overflowing, the timestamp is
+    # advanced and the counter is reset to a random 42-bit integer with MSB
+    # set to 0.
+
+    def get_counter_and_tail():
+        rand = int.from_bytes(os.urandom(10))
+        # 42-bit counter with MSB set to 0
+        counter = (rand >> 32) & 0x1ff_ffff_ffff
+        # 32-bit random data
+        tail = rand & 0xffff_ffff
+        return counter, tail
+
+    global _last_timestamp_v7
+    global _last_counter_v7
+
+    import time
+    nanoseconds = time.time_ns()
+    timestamp_ms = nanoseconds // 1_000_000
+
+    if _last_timestamp_v7 is None or timestamp_ms > _last_timestamp_v7:
+        counter, tail = get_counter_and_tail()
+    else:
+        if timestamp_ms < _last_timestamp_v7:
+            timestamp_ms = _last_timestamp_v7 + 1
+        # advance the 42-bit counter
+        counter = _last_counter_v7 + 1
+        if counter > 0x3ff_ffff_ffff:
+            timestamp_ms += 1  # advance the 48-bit timestamp
+            counter, tail = get_counter_and_tail()
+        else:
+            tail = int.from_bytes(os.urandom(4))
```

Comment:
You might add a comment to add the conversion from bytes (4) to bits (32):

```suggestion
            # 32-bit random data
            tail = int.from_bytes(os.urandom(4))
```